src/network/WiFiHandler.cpp

```diff
@@ -0,0 +1,229 @@
+            }
+        }
+
+        void WiFiHandler::onConnect()
+        {
+            this->logger.info("Connected to \"%s\": %s", WiFi.SSID().c_str(),
+                              WiFi.localIP().toString().c_str());
+
+            statusManager.setStatus(SlimeVR::Status::WIFI_CONNECTED, true);
+            statusManager.setStatus(SlimeVR::Status::WIFI_CONNECTING, false);
+
+            this->connected = true;
+            this->hadWiFi = true;
+
+            this->stopProvisioning();
+        }
+
+        void WiFiHandler::onDisconnect()
+        {
+            this->logger.warn("Connection to WiFi lost, reconnecting...");
+
+            statusManager.setStatus(SlimeVR::Status::WIFI_CONNECTED, false);
+            statusManager.setStatus(SlimeVR::Status::WIFI_CONNECTING, true);
+
+            this->connected = false;
+            this->millisSinceConnectionStart = millis();
+        }
+
+        void WiFiHandler::update()
+        {
+            wl_status_t status = (wl_status_t)this->wifi.run(10000);
+
+            if (status == WL_CONNECTED)
+            {
+                if (!this->connected)
+                {
+                    this->onConnect();
+                }
+
+                if (this->lastRSSISampleTime + 2000 < millis())
+                {
+                    this->lastRSSISampleTime = millis();
+                    uint8_t rssi = WiFi.RSSI();
+                    ::Network::sendSignalStrength(rssi);
+                }
+
+                return;
+            }
+
+            // Call our `onDisconnect` handler when we were connected previously and now aren't
+            // anymore
+            if (this->connected)
+            {
+                this->onDisconnect();
+            }
+
+            if (this->millisSinceConnectionStart + 5000 < millis() && !this->hadWiFi &&
+                !WiFi.smartConfigDone())
+            {
+                if (status == WL_IDLE_STATUS)
+                {
+                    this->logger.error("Can't connect from any credentials, status: %d.",
+                                        WiFi.status());
+                }
+
+                this->startProvisioning();
+            }
+        }
+
+        void WiFiHandler::updateProvisioning()
+        {
+            // Called even when not provisioning to do things like provide neighbours or other
+            // upkeep
+        }
+
+        void WiFiHandler::startProvisioning()
+        {
+            if (!WiFi.beginSmartConfig())
+            {
+                this->logger.error("Could not start SmartConfig!");
+
+                return;
+            }
+
+            this->provisioning = true;
+            this->logger.info("SmartConfig started");
+        }
+
+        void WiFiHandler::stopProvisioning()
+        {
+            WiFi.stopSmartConfig();
+            this->provisioning = false;
+        }
+
+        void WiFiHandler::provideNeighbours()
+        {
+            // TODO: SmartConfig can't do this, created for future
```

Comment:
if it can't do this, remove implementation and comment out declaration in header, mark it as TODO